posts/inside-rust/2021-05-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,235 @@
+
+Finally, we need to specialize `expect` and `unwrap` to use these new `Error` aware panic interfaces when unwrapping types that implement the `Error` trait. To do this we first need to work around a [soundness issue](https://smallcultfollowing.com/babysteps/blog/2018/02/09/maximally-minimal-specialization-always-applicable-impls/#the-soundness-problem) with specialization for trait impls that are conditional based on lifetimes, though thankfully we already have a good idea of [how to fix this](https://smallcultfollowing.com/babysteps/blog/2018/02/09/maximally-minimal-specialization-always-applicable-impls).
+
+### 2. Error Reporter
+
+We would also like to provide a basic error reporter in `std`, and some facilities for making it easy to use, or easy to replace with your own preferred error reporter. Printing an error and its sources is a fundamental operation in Rust, so we want the language to provide a pit of success for reporting, where the easiest thing to do is the right thing. We can't get there completely because we use `Display` for individual error messages, and we can't change that in a backwards compatible fashion, but we hope that adding a convenient method for printing a full chain of errors and some clever lints will relieve most of the pressure.
+
+We plan on fixing this by first adding a `Report` type to the standard library that wraps a `&dyn Error` and implements `Display` such that it prints each source as desired. We would like the output of `Report`'s display method to support the styles of error concatenation that are most common in the rust ecosystem.
+
+Either one line with each error message concatenated with colons:
+
+```rust
+println!("Error: {}", Report::from(error));
+
+// Outputs:
+// Error: outermost error: second error: root error
+```
+
+Or multiple lines with each error message on it's own line :
+
+
+```rust
+println!("Error: {:#}", Report::from(error))
+
+// Outputs:
+// Error: outermost error
+//
+// Caused by:
+//    0: second error
+//    1: root error
+```
+
+The first single line format is useful for log output or inlined error messages, whereas the alternate multi line format is useful for user facing output such as a CLI interface or a GUI popup.
+
+We also want to add a method to the error trait for conveniently wrapping up any error in the `Report` type so that reporting an error is as simple as `println!("Error: {}", error.report());`
+
+We expect the report method will look something like this:
+
+```rust
+fn report(&self) -> impl Display + '_
+where
+    Self: Sized,
+{
+    Report::from(self)
+}
+```
+
+We want the return type here to be generic rather than hard coded to `Report` so that individual error types can provide their own report format if desired. We expect there will be some fun things derive macros can do here to customize error reporting formats. This will work well with composition because the reporter from the outermost type will be used to format the full chain of errors.
```

Comment:
replace
> We expect there will be some fun things derive macros can do here to customize error reporting formats.

with
>We expect there will be some fun things, that derive macros can do here to customize error reporting formats.

?

Reply:
I don't think this is grammatically correct in English.